
SAMPLES17 =['2016_SNC1_A', '2016_SNC1_B', '2016_SNC1_C', 
            '2016_SNC1_D', '2016_RRV3_A', '2016_RRV3_B', 
            '2016_RRV3_C', '2016_RRV3_D', '2016_RRV2_A', 
            '2016_RRV2_B', '2016_RRV2_C', '2016_RRV2_D', 
            '2016_AV1_A', '2016_AV1_B', '2016_AV1_C', 
            '2016_AV1_D', '2016_CRN1_A', '2016_CRN1_B', 
            '2016_CRN1_C', '2016_CRN1_D', '2016_SNC2_A', 
            '2016_SNC2_B', '2016_SNC2_C', '2016_SNC2_D', 
            '2016_OR1_A', '2016_OR1_B', '2016_OR1_C', 
            '2016_OR1_D', '2016_SMV1_A', '2016_SMV1_B', 
            '2016_SMV1_C', '2016_SMV1_D', '2016_AS1_A', 
            '2016_AS1_B', '2016_AS1_C', '2016_AS1_D', 
            '2016_AS2_A', '2016_AS2_B', '2016_AS2_C', 
            '2016_AS2_D', '2016_SRH1_A', '2016_SRH1_B', 
            '2016_SRH1_C', '2016_SRH1_D', '2016_SMV2_A', 
            '2016_SMV2_B', '2016_SMV2_C', '2016_SMV2_D', 
            '2016_RRV1_A', '2016_RRV1_B', '2016_RRV1_C', 
            '2016_RRV1_D', '2017_SNC1_A', '2017_SNC1_B', 
            '2017_SNC1_C', '2017_SNC1_D', '2017_RRV3_A', 
            '2017_RRV3_B', '2017_RRV3_C', '2017_RRV3_D', 
            '2017_RRV2_A', '2017_RRV2_B', '2017_RRV2_C', 
            '2017_RRV2_D', '2017_AV1_A', '2017_AV1_B', 
            '2017_AV1_C', '2017_AV1_D', '2017_CRN1_A', 
            '2017_CRN1_B', '2017_CRN1_C', '2017_CRN1_D', 
            '2017_SNC2_A', '2017_SNC2_B', '2017_SNC2_C', 
            '2017_SNC2_D', '2017_AV2_A', '2017_AV2_B', 
            '2017_AV2_C', '2017_AV2_D', '2017_OR1_A', 
            '2017_OR1_B', '2017_OR1_C', '2017_OR1_D', 
            '2017_AS2_A', '2017_AS2_B', '2017_AS2_C', 
            '2017_AS2_D', '2017_SMV1_A', '2017_SMV1_B', 
            '2017_SMV1_C', '2017_SMV1_D', '2017_AS1_A', 
            '2017_AS1_B', '2017_AS1_C', '2017_AS1_D', 
            '2017_SRH1_A', '2017_SRH1_B', '2017_SRH1_C', 
            '2017_SRH1_D', '2017_SMV2_A', '2017_SMV2_B', 
            '2017_SMV2_C', '2017_SMV2_D', '2017_RRV1_A', 
            '2017_RRV1_B', '2017_RRV1_C', '2017_RRV1_D', '2017_OR2_A', 
            '2017_OR2_B', '2017_OR2_C', '2017_OR2_D',
            '16mockcomm1a', '16mockcomm2a', '16mockcomm3a', 
            '16negctrl1a', '16negctrl2a', '16negctrl3a', 
            '17mockcomm1b', '17mockcomm2b', '17mockcomm3b', 
            '17negctrl1b', '17negctrl2b', '17negctrl3b'] 

SAMPLES19 = ['2019_SNC1_A', '2019_SNC1_B', '2019_SNC1_C', '2019_SNC1_D', 
             '2019_RRV3_A', '2019_RRV3_B', '2019_RRV3_C', '2019_RRV3_D', 
             '2019_RRV2_A', '2019_RRV2_B', '2019_RRV2_C', '2019_RRV2_D', 
             '2019_AV1_A', '2019_AV1_B', '2019_AV1_C', '2019_AV1_D', 
             '2019_CRN1_A', '2019_CRN1_B', '2019_CRN1_C', '2019_CRN1_D', 
             '2019_AV2_A', '2019_AV2_B', '2019_AV2_C', 
             '2019_OR1_A', '2019_OR1_B', '2019_OR1_C', '2019_OR1_D', 
             '2019_AS2_A', '2019_AS2_B', '2019_AS2_C', '2019_AS2_D', 
             '2019_SMV1_A', '2019_SMV1_B', '2019_SMV1_C', '2019_SMV1_D', 
             '2019_AS1_A', '2019_AS1_B', '2019_AS1_C', '2019_AS1_D', 
             '2019_SRH1_A', '2019_SRH1_B', '2019_SRH1_C', '2019_SRH1_D', 
             '2019_SMV2_A', '2019_SMV2_B', '2019_SMV2_C', '2019_SMV2_D', 
             '2019_OR2_A', '2019_OR2_B', '2019_OR2_C', '2019_OR2_D',
             '19CH-50FOO', '19CH-50FPO', '19CH-50FNH', '19CH-70FOO', 
             'negctrl1', 'negctrl2', 'negctrl3', 'negctrl4', 
             'JFW_19-1-2', 'JFW_19-2-2', 'JFW_19-19-2', 'JFW_19-24-2', 
             'JFW_19-19-5', 'JFW_19-20-5', 'JFW_19-23-5', 'JFW_19-24-5']

SAMPLES = SAMPLES17 + SAMPLES19

rule all:
    input:
        "outputs/16s/dada2/summary_table.tsv",
        #"outputs/ITS/dada2/ASV_counts.tsv",

rule make_barcodes_2019_16s:
    input: primers = "inputs/2019_16s_samples_ava.csv"
    output: barcodes = "inputs/2019_16s_barcodes.fa"
    script: "scripts/make_barcodes_sabre.R"

rule demultiplex_2019_16s:
    input:
        barcode = "inputs/2019_16s_barcodes.fa",
        R1 = "inputs/raw/2019PN16S_S2_L001_R1_001.fastq.gz",
        R2 = "inputs/raw/2019PN16S_S2_L001_R2_001.fastq.gz"
    output: 
        R1 = expand("outputs/16s/demultiplexed_19/{sample19}_R1.fq", sample19 = SAMPLES19),
        R2 = expand("outputs/16s/demultiplexed_19/{sample19}_R2.fq", sample19 = SAMPLES19),
        n1 = "outputs/16s/demultiplexed_19/no_match_R1.fq",
        n2 = "outputs/16s/demultiplexed_19/no_match_R2.fq"
    conda: "envs/sabre.yml"
    shell:'''
    # cutadapt -e 0.15 --no-indels -g file:{input.barcode} -o outputs/demultiplexed_19/{{name}}_R1_dm.fq.gz -p outputs/demultiplexed_19/{{name}}_R2_dm.fq.gz {input.R1} {input.R2}
    sabre pe -m 1 -f {input.R1} -r {input.R2} -b {input.barcode} -u {output.n1} -w {output.n2}
    mv *fq outputs/16s/demultiplexed_19/
    '''

rule adapter_trim_2019_16s:
    input:
        R1 = "outputs/16s/demultiplexed_19/{sample19}_R1.fq",
        R2 = "outputs/16s/demultiplexed_19/{sample19}_R2.fq"
    output:
        R1 = "outputs/16s/trimmed_19/{sample19}_R1_trimmed.fq.gz",
        R2 = "outputs/16s/trimmed_19/{sample19}_R2_trimmed.fq.gz"
    conda: "envs/cutadapt.yml"
    shell:'''
    cutadapt -g GTGCCAGCMGCCGCGGTAA -a ATTAGAWACCCBNGTAGTCC \
             -G GGACTACNVGGGTWTCTAAT -A TTACCGCGGCKGCTGGCAC \
             -m 160 -M 235 -o {output.R1} -p {output.R2} {input.R1} {input.R2} \
             >> cutadapt_primer_trimming_stats.txt
    '''

rule make_barcodes_2017_16s:
    input: primers = "inputs/2017_2016_16s_samples_ava.csv"
    output: barcodes = "inputs/2017_16s_barcodes.fa"
    script: "scripts/make_barcodes_sabre.R"

rule demultiplex_2017_16s:
    input:
        barcode = "inputs/2017_16s_barcodes.fa",
        R1 = "inputs/raw/16S-16-17-Must_S1_L001_R1_001.fastq.gz",
        R2 = "inputs/raw/16S-16-17-Must_S1_L001_R2_001.fastq.gz"
    output: 
        R1 = expand("outputs/16s/demultiplexed_17/{sample17}_R1.fq", sample17 = SAMPLES17),
        R2 = expand("outputs/16s/demultiplexed_17/{sample17}_R2.fq", sample17 = SAMPLES17),
        n1 = "outputs/16s/demultiplexed_17/no_match_R1.fq",
        n2 = "outputs/16s/demultiplexed_17/no_match_R2.fq"
    conda: "envs/sabre.yml"
    shell:'''
    #cutadapt -e 0.15 --no-indels -g file:{input.barcode} -o outputs/demultiplexed_17/{{name}}_R1_dm.fq.gz -p outputs/demultiplexed_17/{{name}}_R2_dm.fq.gz {input.R1} {input.R2}
    sabre pe -m 1 -f {input.R1} -r {input.R2} -b {input.barcode} -u {output.n1} -w {output.n2}
    mv *fq outputs/16s/demultiplexed_17/
    '''

rule adapter_trim_2017_16s:
    input:
        R1 = "outputs/16s/demultiplexed_17/{sample17}_R1.fq",
        R2 = "outputs/16s/demultiplexed_17/{sample17}_R2.fq"
    output:
        R1 = "outputs/16s/trimmed_17/{sample17}_R1_trimmed.fq.gz",
        R2 = "outputs/16s/trimmed_17/{sample17}_R2_trimmed.fq.gz"
    conda: "envs/cutadapt.yml"
    shell:'''
    cutadapt -j 3 -g GTGCCAGCMGCCGCGGTAA -a ATTAGAWACCCBNGTAGTCC \
             -G GGACTACNVGGGTWTCTAAT -A TTACCGCGGCKGCTGGCAC \
             -m 160 -M 235 -o {output.R1} -p {output.R2} {input.R1} {input.R2} \
             >> 2017_2016_16s_cutadapt_primer_trimming_stats.txt
    '''

rule download_silva:
    output: "inputs/silva_nr_v138_train_set.fa.gz"
    conda: "envs/wget.yml"
    shell:'''
    wget -O {output} https://zenodo.org/record/3731176/files/silva_nr_v138_train_set.fa.gz?download=1
    '''

rule dada2:
    input: 
        expand("outputs/16s/trimmed_17/{sample17}_R1_trimmed.fq.gz", sample17 = SAMPLES17),
        expand("outputs/16s/trimmed_17/{sample17}_R2_trimmed.fq.gz", sample17 = SAMPLES17),
        expand("outputs/16s/trimmed_19/{sample19}_R1_trimmed.fq.gz", sample19 = SAMPLES19),
        expand("outputs/16s/trimmed_19/{sample19}_R2_trimmed.fq.gz", sample19 = SAMPLES19),
        silva = "inputs/silva_nr_v138_train_set.fa.gz",
    output:
        #quality_plts = "outputs/16s/dada2/quality_plots.pdf",
        error_plts = "outputs/16s/dada2/error_plots.pdf",
        asvs = "outputs/16s/dada2/ASVs.fa",
        counts = "outputs/16s/dada2/ASV_counts.tsv",
        tax = "outputs/16s/dada2/ASV_taxonomy.tsv",
        summary = "outputs/16s/dada2/summary_table.tsv"
    conda: "envs/dada2.yml"
    script: "scripts/16s_dada2.R"

################################################
### ITS sequencing
################################################

rule make_barcodes_2019_ITS:
    input: primers = "inputs/2019_ITS_samples_ava.csv"
    output: barcodes = "inputs/2019_ITS_barcodes.tsv"
    script: "scripts/make_barcodes_sabre.R"

rule demultiplex_2019_ITS:
    input:
        barcode = "inputs/2019_ITS_barcodes.tsv",
        R1 = "inputs/raw/2019PNBITS_S1_L001_R1_001.fastq.gz",
        R2 = "inputs/raw/2019PNBITS_S1_L001_R2_001.fastq.gz"
    output: 
        R1 = expand("outputs/ITS/demultiplexed_19/{sample19}_R1.fq", sample19 = SAMPLES19),
        R2 = expand("outputs/ITS/demultiplexed_19/{sample19}_R2.fq", sample19 = SAMPLES19),
        n1 = "outputs/ITS/demultiplexed_19/no_match_R1.fq",
        n2 = "outputs/ITS/demultiplexed_19/no_match_R2.fq"
    conda: "envs/sabre.yml"
    shell:'''
    sabre pe -m 1 -f {input.R1} -r {input.R2} -b {input.barcode} -u {output.n1} -w {output.n2}
    mv *fq outputs/ITS/demultiplexed_19/
   '''

rule adapter_trim_2019_ITS:
    input:
        R1 = "outputs/ITS/demultiplexed_19/{sample19}_R1.fq",
        R2 = "outputs/ITS/demultiplexed_19/{sample19}_R2.fq"
    output:
        R1 = "outputs/ITS/trimmed_19/{sample19}_R1_trimmed.fq",
        R2 = "outputs/ITS/trimmed_19/{sample19}_R2_trimmed.fq"
    conda: "envs/cutadapt.yml"
    shell:'''
    cutadapt -g CTACCTGCGGARGGATCA -a AAACTTTYARCAAYGGATCTC \
             -G GAGATCCRTTGYTRAAAGTTT -A TGATCCYTCCGCAGGTAG \
             -m 160 -M 240 -o {output.R1} -p {output.R2} {input.R1} {input.R2} \
             >> cutadapt_primer_trimming_stats_ITS_2019.txt
    '''

rule make_barcodes_2017_ITS:
    input: primers = "inputs/2017_2016_ITS_samples_ava.csv"
    output: barcodes = "inputs/2017_2016_ITS_barcodes.tsv"
    script: "scripts/make_barcodes_sabre.R"

rule demultiplex_2017_ITS:
    input:
        barcode = "inputs/2017_2016_ITS_barcodes.tsv",
        R1 = "inputs/raw/BITS-16-17-Must_S1_L001_R1_001.fastq.gz",
        R2 = "inputs/raw/BITS-16-17-Must_S1_L001_R2_001.fastq.gz"
    output: 
        R1 = expand("outputs/ITS/demultiplexed_17/{sample17}_R1.fq", sample17 = SAMPLES17),
        R2 = expand("outputs/ITS/demultiplexed_17/{sample17}_R2.fq", sample17 = SAMPLES17),
        n1 = "outputs/ITS/demultiplexed_17/no_match_R1.fq",
        n2 = "outputs/ITS/demultiplexed_17/no_match_R2.fq"
    conda: "envs/sabre.yml"
    shell:'''
    sabre pe -m 1 -f {input.R1} -r {input.R2} -b {input.barcode} -u {output.n1} -w {output.n2}
    mv *fq outputs/ITS/demultiplexed_17/
    '''

rule adapter_trim_2017_ITS:
    input:
        R1 = "outputs/ITS/demultiplexed_17/{sample17}_R1.fq",
        R2 = "outputs/ITS/demultiplexed_17/{sample17}_R2.fq"
    output:
        R1 = "outputs/ITS/trimmed_17/{sample17}_R1_trimmed.fq",
        R2 = "outputs/ITS/trimmed_17/{sample17}_R2_trimmed.fq"
    conda: "envs/cutadapt.yml"
    shell:'''
    cutadapt -g CTACCTGCGGARGGATCA -a AAACTTTYARCAAYGGATCTC \
             -G GAGATCCRTTGYTRAAAGTTT -A TGATCCYTCCGCAGGTAG \
             -m 160 -M 240 -o {output.R1} -p {output.R2} {input.R1} {input.R2} \
             >> cutadapt_primer_trimming_stats_ITS_2017.txt
    '''

rule download_unite:
    output: "inputs/sh_general_release_s_04.02.2020.tar.gz"
    conda: "envs/wget.yml"
    shell:'''
    wget -O {output} https://files.plutof.ut.ee/public/orig/1E/66/1E662B6EB320312A61E7E3218327F34C7DB09CFF8E4686A89EF47886822DA6AB.gz
    '''

rule untar_unite:
    input: "inputs/sh_general_release_s_04.02.2020.tar.gz"
    output: 'inputs/sh_general_release_s_04.02.2020/sh_general_release_dynamic_s_04.02.2020.fasta'
    shell:'''
    tar xf {input} -C inputs/    
    '''

rule dada2_ITS:
    input: 
        expand("outputs/ITS/trimmed_17/{sample17}_R1_trimmed.fq", sample17 = SAMPLES17),
        expand("outputs/ITS/trimmed_17/{sample17}_R2_trimmed.fq", sample17 = SAMPLES17),
        expand("outputs/ITS/trimmed_19/{sample19}_R1_trimmed.fq", sample19 = SAMPLES19),
        expand("outputs/ITS/trimmed_19/{sample19}_R2_trimmed.fq", sample19 = SAMPLES19),
        unite = "inputs/sh_general_release_s_04.02.2020/sh_general_release_dynamic_s_04.02.2020.fasta",
    output:
        quality_plts = "outputs/ITS/dada2/quality_plots.pdf",
        error_plts = "outputs/ITS/dada2/error_plots.pdf",
        asvs = "outputs/ITS/dada2/ASVs.fa",
        counts = "outputs/ITS/dada2/ASV_counts.tsv",
        tax = "outputs/ITS/dada2/ASV_taxonomy.tsv",
        summary = "outputs/ITS/dada2/summary_table.tsv"
    conda: "envs/dada2.yml"
    script: "scripts/ITS_dada2.R"
